python-frontend/src/main/java/org/sonar/python/tree/CallExpressionImpl.java

```diff
@@ -187,13 +188,23 @@ public List<Expression> typeDependencies() {
 
   @Override
   public PythonType typeV2() {
-    if (callee().typeV2() instanceof ClassType classType) {
-      return new ObjectType(classType);
+    TypeSource typeSource = computeTypeSource();
+    PythonType pythonType = returnTypeOfCall(callee().typeV2());
+    return pythonType != PythonType.UNKNOWN ? new ObjectType(pythonType, typeSource) : PythonType.UNKNOWN;
+  }
+
+  static PythonType returnTypeOfCall(PythonType calleeType) {
+    if (calleeType instanceof ClassType classType) {
+      return classType;
     }
-    if (callee().typeV2() instanceof FunctionType functionType) {
-      return functionType.returnType();
+    if (calleeType instanceof FunctionType functionType) {
+      PythonType returnType = functionType.returnType();
+      if (returnType.equals(PythonType.UNKNOWN)) {
+        return PythonType.UNKNOWN;
+      }
+      return returnType;
     }
-    if (callee().typeV2() instanceof UnionType unionType) {
+    if (calleeType instanceof UnionType unionType) {
```

Comment:
I guess it makes sense to make union-type case processing recursively - get the return type of every candidate and join them over the `UnionType.or` method

Reply:
`UnionType.or` itself should already be recursive. However, I made some slight changes to make this piece of code more intentional.
In particular, if the original `UnionType` contains anything unexpected, I think we should return `PythonType.UNKNOWN`.